src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BasePatternSwitchLocalRewriter.cs

```diff
@@ -126,95 +125,171 @@ protected virtual LabelSymbol GetDagNodeLabel(BoundDecisionDag dag)
             /// <summary>
             /// Lower the given nodes into _loweredDecisionDag.
             /// </summary>
-            private void LowerDecisionDag(ImmutableArray<BoundDecisionDag> sortedNodes)
+            private void LowerDecisionDag(ImmutableArray<BoundDecisionDagNode> sortedNodes)
             {
                 switch (sortedNodes[0])
                 {
-                    case BoundWhenClause wc:
+                    case BoundWhenDecisionDagNode wc:
                         // If the first node is in a when clause's section rather than the code for the
                         // lowered decision dag, jump there to start.
                         _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(wc)));
                         break;
-                    case BoundDecision d:
-                        // If the first node is a decision rather than the code for the
+                    case BoundLeafDecisionDagNode d:
+                        // If the first node is a leaf rather than the code for the
                         // lowered decision dag, jump there to start.
                         _loweredDecisionDag.Add(_factory.Goto(GetDagNodeLabel(d)));
                         break;
                 }
 
-                // Call LowerDecisionDagNode with each node and its following node in the generation order.
-                // However, some nodes can be emitted more efficiently as a switch dispatch (possibly out
-                // of order).
-                var loweredNodes = PooledHashSet<BoundDecisionDag>.GetInstance();
-                BoundDecisionDag previous = null;
-                foreach (BoundDecisionDag node in sortedNodes)
+                // Note that a when-clause can contain an assignment to a
+                // pattern variable declared in a different when-clause, so we only apply the optimization
+                // of using user-declared pattern variables as pattern-matching automaton temps
+                // when there are no nontrivial when-clauses at all.
+                if (!sortedNodes.OfType<BoundWhenDecisionDagNode>().Where(w => w.WhenExpression != null && w.WhenExpression.ConstantValue != ConstantValue.True).Any())
```

Comment:
>sortedNodes [](start = 21, length = 11)

Consider using `Any` only, to avoid allocations: `sortedNodes.Any(n => { var w = n as BoundWhenDecisionDagNode; return w.WhenExpression != null && ...; })`